sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/statefulOperators.scala

```diff
@@ -96,6 +98,25 @@ trait StateStoreReader extends StatefulOperator {
 /** An operator that writes to a StateStore. */
 trait StateStoreWriter extends StatefulOperator with PythonSQLMetrics { self: SparkPlan =>
 
+  /**
+   * Produce the output watermark for given input watermark (ms).
+   *
+   * In most cases, this is same as the criteria of state eviction, as most stateful operators
+   * produce the output from two different kinds:
+   *
+   * 1. without buffering
+   * 2. with buffering (state)
+   *
+   * The state eviction happens when event time exceeds a "certain threshold of timestamp", which
+   * denotes a lower bound of event time values for output (output watermark).
+   *
+   * The default implementation provides the input watermark as it is. Most built-in operators
+   * will evict based on min input watermark and ensure it will be minimum of the event time value
+   * for the output so far (including output from eviction). Operators which behave differently
+   * (e.g. different criteria on eviction) must override this method.
+   */
+  def produceOutputWatermark(inputWatermarkMs: Long): Option[Long] = Some(inputWatermarkMs)
```

Comment:
Why is this the default behavior for stateful operators?  Consider a query with two window operators:

source -> 5 minute window - > ten min window

Batch 0:

Input data: 1, 2, 3

5 minute window operator watermark should be 3 after batch 0 but the watermark of the ten minute window should still be 0.

Reply:
Your example is not consistent. If you want to advance watermark "after" you produce output, both 5 minutes and 10 minutes window operator watermark should be 0.

Even we provide the output watermark as same as input watermark, it does not violate the definition of watermark, lower bound of the "output". Advancing watermark as 3 for 5 minutes window does not break the logic. Further output will be at least 5, not less than 3, hence the previous advancement of watermark is never violated. Current watermark logic has been working like this way.

Reply:
The only case this could become wrong is something like source -> 5 minutes window -> 3 minutes window, but can someone ever infer the right behavior and output for this case? Is it semantically making sense? I don't think so, so I just excluded the case.

Reply:
So given the current logic: 

wm1 = input watermark for 5 minute window
wm2 = input watermark for 10 minute window

Batch 0: wm1 = 0 wm2 = 0
batch 1: wm1 = 3 wm2 = 3

While this strategy may work in the current implementation and for most use cases,  is it comprehensive?  This works in situations where you have have windowed operators with increasing window sizes but as you pointed out what if that is not true.  Like in the example you provided source -> 5 minutes window -> 3 minutes window, the 3 minute window would output a result before the 5 minute window would which is incorrect.   The example you provided if the 5 minute window operator outputted a window of 0-5, this window record should be buffered as an entry in 3-6 window range for the 3 minute window.  I don't have a specific use case in mind of why someone would do this but why shouldn't it be possible?

Reply:
While we say "representative time" of time window, we still need to reason about the semantic for chained time window aggregation, based on how time windows are "inclusive" in bigger time window.

For example, if we suppose to have a query having 5 mins time window -> 3 mins time window. Are we really sure having any output in 3 mins time window is valid? Where the output of 5 mins time window have to bound for next 3 mins time window? Why we need to worry about the case which just doesn't make sense at all?

Reply:
And... even though the semantic is simply wrong, 0-5 window will be buffered in 3-6 window range, even for this implementation. Let's revisit your example:

batch 1: wm1 = 3 wm2 = 3. wm2 may try to evict the buffered window but nothing is there.

and data 4, 5, 6, 7 has been ingested.

batch 2: wm1 = 7 wm2 = 7, 0-5 window is produced from op1 which is NOT a late record for wm2 (it was 3), hence processed, and then wm2 for eviction (7) is applied and 3-6 window is produced from op2.

Reply:
@HeartSaVioR thanks for the detailed explanation.  Can you add we add this information as a comment in the code so we don't forget the rationale of why we are doing this?

Reply:
OK. I'll leave a context as a code comment. I agree this is missing and leads some confusion.

Reply:
[c7946f7](https://github.com/apache/spark/pull/39931/commits/c7946f7d6718a9866e97dfcb262243bfb14e0b5a)